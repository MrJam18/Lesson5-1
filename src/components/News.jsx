import React from 'react';
import { useState } from 'react';
import { useEffect } from 'react';
import styles from '../css/news.module.css';
import '../css/App.css'
import { getError, getLoading, getNews } from '../store/news/selectors';
import { fetchNews } from '../store/news/actions';
import {useSelector, useDispatch} from 'react-redux';
import { CircularProgress } from '@material-ui/core';
 let newsID = 0;

const News = () => {
    const news = useSelector(getNews);
    const err = useSelector(getError);
    const loading = useSelector(getLoading);
    const dispatch = useDispatch();

    const ErrorComp = () => {
        return (
            <div className={styles.errorBox}>
            <div className={styles.error}>Error! {err}. </div>
            <button onClick={()=> dispatch(fetchNews())}>Retry</button>
            </div>
        )
    }
    const Content = () => news.map((el)=> {
        newsID++;
        return(
        <a href={el.url} className={styles.newsBlock} key = {newsID}>
        <div className={'small-header ' + styles.boldText}>{el.title}</div>
               {el.image != null && <img src={el.image} alt={el.source} className={styles.image}/>}
               <div className={styles.text}>{el.description}</div>
            <div className={styles.source}>Источник: {el.source}.</div>
        </a>
        )
    })

    useEffect(() => dispatch(fetchNews()), [])
    
    return (
        <div className={styles.news}>
        {loading && <div className={styles.loading}><CircularProgress color="secondary" size= '200px'/></div>}
        {err ? <ErrorComp/> : <Content/>}
        </div>
    );
};

export default News;