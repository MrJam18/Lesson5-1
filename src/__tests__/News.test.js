import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import News from "../components/News";
import configureStore from 'redux-mock-store' //ES6 modules
import thunk from "redux-thunk";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initStore = { news:{
    news: [
      {
        author: 'Ксения Назарова',
        title: 'Бывший владелец Virtus.pro выиграл 44 миллиона рублей на победе Team Spirit',
        description: 'Бывший владелец российской киберспортивной команды Virtus.pro Антон Черепенников выиграл более 44 миллионов рублей после победы Team Spirit в финале международных соревнований по Dota 2 The International 10. Бизнесмен поставил 1,53 миллиона рублей при коэффициенте ставки 29. «Верил с 1 дня», — написал он.',
        url: 'https://lenta.ru/news/2021/10/18/44_team_spirit/',
        source: 'Lenta',
        image: 'https://icdn.lenta.ru/images/2021/10/18/05/20211018050423260/pic_19538ff4b39d63fe7d0cea14f74efd7e.jpg',
        category: 'general',
        language: 'ru',
        country: 'ru',
        published_at: '2021-10-18T02:17:18+00:00'
      },
      {
        author: null,
        title: 'Hyundai Tucson и Toyota Yaris Cross стали отличниками краш-тестов',
        description: 'Hyundai Tucson и Toyota Yaris Cross стали отличниками краш-тестов',
        url: 'https://rg.ru/2021/10/27/hyundai-tucson-i-toyota-yaris-cross-stali-otlichnikami-krash-testov.html',
        source: 'RG.ru Rossiyskaya Gazeta',
        image: 'https://cdnimg.rg.ru/img/content/219/79/15/596562.jpg',
        category: 'general',
        language: 'ru',
        country: 'ru',
        published_at: '2021-10-27T14:02:05+00:00'
      },
      {
        author: null,
        title: 'В РКН рассказали о штрафах Facebook, Twitter, Telegram, Google и Tiktok',
        description: 'Российские суды в 2021 году за неудаление запрещенной информации оштрафовали Facebook, Twitter, Telegram, Google и TikTok на 180 миллионов рублей, сообщил Роскомнадзор.',
        url: 'https://ria.ru/20211101/shtraf-1757182301.html',
        source: 'RIA Novosti',
        image: null,
        category: 'general',
        language: 'ru',
        country: 'ru',
        published_at: '2021-11-01T09:31:37+00:00'
      },
      {
        author: 'Марина Совина',
        title: 'Российско-украинская команда NaVi выиграла чемпионат мира по CS:GO',
        description: 'Команда NaVi, в состав которой входят российские и украинские киберспортсмены, победила в турнире PGL Major Stockholm 2021 по компьютерной игре CS:GO. В финале NaVi обыграли европейскую команду G2 Esports. Встреча завершилась со счетом 2-0 (16:11 на Ancient, 22:19 на Nuke).',
        url: 'https://lenta.ru/news/2021/11/08/csgo/',
        source: 'Lenta.ru',
        image: 'https://icdn.lenta.ru/images/2021/11/08/01/20211108012447548/pic_f8fd2b79d9882f42236494807bb099f4.jpg',
        category: 'general',
        language: 'ru',
        country: 'ru',
        published_at: '2021-11-07T22:40:01+00:00'
      }
    ]
}
}

it('test News to compare last action with expected', ()=> {
    const store = mockStore(initStore);
    render(
        <Provider store={store}>
          <News />
        </Provider>
      );
    
      const actions = store.getActions();
      const lastAction = actions[actions.length - 1];
      const expected = {
        type: 'NEWS::SET_LOADING',
        loading: true
    }
      expect(lastAction).toEqual(expected);
})
it('test News to Compare render with snapshot', ()=> {
    const store = mockStore(initStore);
    const component = render(
        <Provider store={store}>
          <News />
        </Provider>
      );
    
    expect(component).toMatchSnapshot();
})