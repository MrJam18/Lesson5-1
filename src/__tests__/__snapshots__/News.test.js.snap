// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test News to Compare render with snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="news"
      >
        <a
          class="newsBlock"
          href="https://lenta.ru/news/2021/10/18/44_team_spirit/"
        >
          <div
            class="small-header boldText"
          >
            Бывший владелец Virtus.pro выиграл 44 миллиона рублей на победе Team Spirit
          </div>
          <img
            alt="Lenta"
            class="image"
            src="https://icdn.lenta.ru/images/2021/10/18/05/20211018050423260/pic_19538ff4b39d63fe7d0cea14f74efd7e.jpg"
          />
          <div
            class="text"
          >
            Бывший владелец российской киберспортивной команды Virtus.pro Антон Черепенников выиграл более 44 миллионов рублей после победы Team Spirit в финале международных соревнований по Dota 2 The International 10. Бизнесмен поставил 1,53 миллиона рублей при коэффициенте ставки 29. «Верил с 1 дня», — написал он.
          </div>
          <div
            class="source"
          >
            Источник: 
            Lenta
            .
          </div>
        </a>
        <a
          class="newsBlock"
          href="https://rg.ru/2021/10/27/hyundai-tucson-i-toyota-yaris-cross-stali-otlichnikami-krash-testov.html"
        >
          <div
            class="small-header boldText"
          >
            Hyundai Tucson и Toyota Yaris Cross стали отличниками краш-тестов
          </div>
          <img
            alt="RG.ru Rossiyskaya Gazeta"
            class="image"
            src="https://cdnimg.rg.ru/img/content/219/79/15/596562.jpg"
          />
          <div
            class="text"
          >
            Hyundai Tucson и Toyota Yaris Cross стали отличниками краш-тестов
          </div>
          <div
            class="source"
          >
            Источник: 
            RG.ru Rossiyskaya Gazeta
            .
          </div>
        </a>
        <a
          class="newsBlock"
          href="https://ria.ru/20211101/shtraf-1757182301.html"
        >
          <div
            class="small-header boldText"
          >
            В РКН рассказали о штрафах Facebook, Twitter, Telegram, Google и Tiktok
          </div>
          <div
            class="text"
          >
            Российские суды в 2021 году за неудаление запрещенной информации оштрафовали Facebook, Twitter, Telegram, Google и TikTok на 180 миллионов рублей, сообщил Роскомнадзор.
          </div>
          <div
            class="source"
          >
            Источник: 
            RIA Novosti
            .
          </div>
        </a>
        <a
          class="newsBlock"
          href="https://lenta.ru/news/2021/11/08/csgo/"
        >
          <div
            class="small-header boldText"
          >
            Российско-украинская команда NaVi выиграла чемпионат мира по CS:GO
          </div>
          <img
            alt="Lenta.ru"
            class="image"
            src="https://icdn.lenta.ru/images/2021/11/08/01/20211108012447548/pic_f8fd2b79d9882f42236494807bb099f4.jpg"
          />
          <div
            class="text"
          >
            Команда NaVi, в состав которой входят российские и украинские киберспортсмены, победила в турнире PGL Major Stockholm 2021 по компьютерной игре CS:GO. В финале NaVi обыграли европейскую команду G2 Esports. Встреча завершилась со счетом 2-0 (16:11 на Ancient, 22:19 на Nuke).
          </div>
          <div
            class="source"
          >
            Источник: 
            Lenta.ru
            .
          </div>
        </a>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="news"
    >
      <a
        class="newsBlock"
        href="https://lenta.ru/news/2021/10/18/44_team_spirit/"
      >
        <div
          class="small-header boldText"
        >
          Бывший владелец Virtus.pro выиграл 44 миллиона рублей на победе Team Spirit
        </div>
        <img
          alt="Lenta"
          class="image"
          src="https://icdn.lenta.ru/images/2021/10/18/05/20211018050423260/pic_19538ff4b39d63fe7d0cea14f74efd7e.jpg"
        />
        <div
          class="text"
        >
          Бывший владелец российской киберспортивной команды Virtus.pro Антон Черепенников выиграл более 44 миллионов рублей после победы Team Spirit в финале международных соревнований по Dota 2 The International 10. Бизнесмен поставил 1,53 миллиона рублей при коэффициенте ставки 29. «Верил с 1 дня», — написал он.
        </div>
        <div
          class="source"
        >
          Источник: 
          Lenta
          .
        </div>
      </a>
      <a
        class="newsBlock"
        href="https://rg.ru/2021/10/27/hyundai-tucson-i-toyota-yaris-cross-stali-otlichnikami-krash-testov.html"
      >
        <div
          class="small-header boldText"
        >
          Hyundai Tucson и Toyota Yaris Cross стали отличниками краш-тестов
        </div>
        <img
          alt="RG.ru Rossiyskaya Gazeta"
          class="image"
          src="https://cdnimg.rg.ru/img/content/219/79/15/596562.jpg"
        />
        <div
          class="text"
        >
          Hyundai Tucson и Toyota Yaris Cross стали отличниками краш-тестов
        </div>
        <div
          class="source"
        >
          Источник: 
          RG.ru Rossiyskaya Gazeta
          .
        </div>
      </a>
      <a
        class="newsBlock"
        href="https://ria.ru/20211101/shtraf-1757182301.html"
      >
        <div
          class="small-header boldText"
        >
          В РКН рассказали о штрафах Facebook, Twitter, Telegram, Google и Tiktok
        </div>
        <div
          class="text"
        >
          Российские суды в 2021 году за неудаление запрещенной информации оштрафовали Facebook, Twitter, Telegram, Google и TikTok на 180 миллионов рублей, сообщил Роскомнадзор.
        </div>
        <div
          class="source"
        >
          Источник: 
          RIA Novosti
          .
        </div>
      </a>
      <a
        class="newsBlock"
        href="https://lenta.ru/news/2021/11/08/csgo/"
      >
        <div
          class="small-header boldText"
        >
          Российско-украинская команда NaVi выиграла чемпионат мира по CS:GO
        </div>
        <img
          alt="Lenta.ru"
          class="image"
          src="https://icdn.lenta.ru/images/2021/11/08/01/20211108012447548/pic_f8fd2b79d9882f42236494807bb099f4.jpg"
        />
        <div
          class="text"
        >
          Команда NaVi, в состав которой входят российские и украинские киберспортсмены, победила в турнире PGL Major Stockholm 2021 по компьютерной игре CS:GO. В финале NaVi обыграли европейскую команду G2 Esports. Встреча завершилась со счетом 2-0 (16:11 на Ancient, 22:19 на Nuke).
        </div>
        <div
          class="source"
        >
          Источник: 
          Lenta.ru
          .
        </div>
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
